core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1513,23 +1542,30 @@ class KafkaController(val config: KafkaConfig,
           currentLeader != preferredReplica
         }
 
-        val electionErrors = onPreferredReplicaElection(electablePartitions, electionType)
-        val successfulPartitions = electablePartitions -- electionErrors.keySet
-        val results = electionErrors.map { case (partition, ex) =>
-          val apiError = if (ex.isInstanceOf[StateChangeFailedException])
-            new ApiError(Errors.PREFERRED_LEADER_NOT_AVAILABLE, ex.getMessage)
-          else
-            ApiError.fromThrowable(ex)
-          partition -> apiError
+        val results = onReplicaElection(electablePartitions, electionType, electionTrigger).mapValues {
```

Comment:
Seems the criteria for `electablePartitions` only accounts for preferred leader election. For unclean leader election, this seems to work since the set of partitions eligible for unclean election is a subset, but maybe it is better to select the replicas explicitly?

Reply:
Yeah. I am hesitant to make this change in this PR. As it would duplicate code in the PartitionStateMachine here and as a result creating technical debt. In the long term, I  want to refactor this code so that all of these decisions are made in the PartitionStateMachine which returns enough information for this layer to either filter those "errors" or to pass them along to the client.

What do you think?

Reply:
I was considering something simple like just collecting all the offline partitions.